### ISSUE TITLE
feat: Add Material Design 3 (Material You) support

### ISSUE THREAD

**Issue by danielkleebinder** [ionic-team/ionic-framework#24282]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already include this feature request, without success.

### Describe the Feature Request

Recently, the new [Material Design 3](https://m3.material.io/) (publicly known as _Material You_) Guidelines were released. They are quite a cut compared to the current [Material Design 2](https://material.io/), but are now considered the new standard on newer Android devices (Android 12+). Google is also updating all their apps to conform their new specs. Can we expect the Ionic Framework to support the new look and feel for Android devices, and if so, are there already any plans with which release we can expect it? Personally, I think Material You is a great improvement over MD2 and it would be nice if we could ship our apps with this new and awesome design as soon as possible.

Guidelines: [https://m3.material.io/](https://m3.material.io/)

### Describe the Use Case

As an Android 12+ user I want all my apps to fit into the new design ecosystem created by Google.

### Describe Preferred Solution

I think it would be nice if the Material You Design is an opt-in feature that developers can toggle themselves (for example during start up time dynamically or compiled into the final bundle). An exemplary configuration example is listed below:

```typescript
@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, IonicModule.forRoot({
    mode: 'ios' | 'md' | 'md3'
  }), AppRoutingModule],
  providers: [{provide: RouteReuseStrategy, useClass: IonicRouteStrategy}],
  bootstrap: [AppComponent],
})
export class AppModule {
}

```

### Describe Alternatives

Provide some helper tools (for example using the CLI or some web tool) that automatically generate a CSS file based on a given primary color that overrides the default Ionic styles to match with the Material You Design Guidelines. Google themselves already created a tool like this for the web for native Android applications: [https://material-foundation.github.io/material-theme-builder/](https://material-foundation.github.io/material-theme-builder/)

**Comment by sean-perkins** [comment]
@danielkleebinder thanks for the thorough write up and links to supporting documentation. 

In our short term, our big focus internally is releasing Ionic 6 and all the supporting material around it (blog posts, docs, issues, etc.). 

Aligning our components to native design standards has always been a continuous goal and Material You will likely fit into a new release of Ionic in the future. 

If everyone could upvote this issue (👍 ) to help let us know it's important to you and your app development efforts, that'd be great!

**Comment by leonardorafael** [comment]
Hello my friends, if you need MATERIAL DESIGN 3 how about https://www.beercss.com?

![theme-1](https://user-images.githubusercontent.com/12979016/149222946-f7158c88-8052-434f-94ac-ce0cfa350f97.gif)

**Comment by miguelcanosantana** [comment]
Whilst that's a really nice library, I haven't been able to make it work (maybe because my level). But would be nice to have something more official.

**Comment by zac404** [comment]
@sean-perkins Are there some news regarding this topic?

**Comment by PAT-STU** [comment]
Dear Ionic Team, we'd also greatly appreciate a solution like the one mentioned above.

**Comment by sean-perkins** [comment]
Hello everyone, at this time we still are actively gathering community interest and feedback surrounding MD3. There are a variety of changes the team is investigating and working on to make Ionic Framework's design more customizable. 

This feature request is currently our second most upvoted feature request. Please keep upvoting to advocate for this feature.

While this feature will not be included in the v7 release of Ionic Framework. The team will be investigating and digesting the scope of MD3 to determine how to best support it.

Noting valuable resources:
- Official Figma design spec for MD3: https://www.figma.com/community/file/1035203688168086460

**Comment by xnousnow** [comment]
<img width="584" alt="image" src="https://github.com/ionic-team/ionic-framework/assets/119923091/4577ff3a-716e-48a9-af13-384b68643d68">

The demo image in the `UIs that feel at home wherever they are deployed` section on the main page is updated. The ionic `ion-tab-bar` component itself is not updated though.

**Comment by extrordinaire** [comment]
> <img alt="image" width="584" src="https://user-images.githubusercontent.com/119923091/238130601-4577ff3a-716e-48a9-af13-384b68643d68.png">
> The demo image in the `UIs that feel at home wherever they are deployed` section on the main page is updated. The ionic `ion-tab-bar` component itself is not updated though.

I see it as the outdated M2 one though... Checked today, hope Ionic moves towards M3 shortly.

<img width="960" alt="image" src="https://github.com/ionic-team/ionic-framework/assets/112337037/d6440d59-1b18-4b35-9a18-3fbfb78665f5">

**Comment by laurentpayot** [comment]
> <img alt="image" width="584" src="https://user-images.githubusercontent.com/119923091/238130601-4577ff3a-716e-48a9-af13-384b68643d68.png">
> The demo image in the `UIs that feel at home wherever they are deployed` section on the main page is updated. The ionic `ion-tab-bar` component itself is not updated though.

Advertising with MD3 snapshots for something that does only MD2 is illegal in my country...
@sean-perkins any news regarding this MD3 advertising?

**Comment by sean-perkins** [comment]
@laurentpayot when developing Ionic applications you are building with web technologies (HTML, JS, CSS). Ionic components can be styled and swapped out with alternatives to build different experiences (we don't restrict you from combining another UI library with your Ionic apps). 

There are community projects that have MD3 tab designs with Ionic Framework today: https://github.com/tanishqmanuja/ng-md3-ionic-playground

Ionic does not include a "graph component" and the screenshot also shows that. The screenshot is just an example of what types of applications developers _can_ build with Ionic and web technologies. 

--

Let's keep the discussion for this feature on what developers would like to see from Material Design 3 in Ionic. The team is actively investigating this support. We value community feedback on which components, features and aspects of MD3 are most important to your applications. Thank you!

**Comment by xnousnow** [comment]
I think we can use HTML and CSS's new page motion feature, but I'm not sure about that feature to work with the router.

**Comment by jaytonic** [comment]
Hi everyone. Is this now considered? Honestly, my team is reluctant to start new app using ionic because it will looks outdated even before we finished it, and I really don't see the point of having every individual trying to adapt ionic to have an up-to-date UI. I mean, material you is not that new, it has not been integrated in ionic 6 and ionic 7.

**Comment by sean-perkins** [comment]
@jaytonic can you help improve the feedback by listing out which components your team is looking to use that look outdated and what features you would like to see from Material You? There is **a lot** to Material You as our team continues to design a document for all the potential component changes and features. The community can help us by voicing which features are the most important to your applications _now_ versus features we may be able to postpone or not do at all. 

@danielkleebinder transitions, animations and motions would be supported through a few of the interactions that are demonstrated on their site. Mainly the page-to-page router transitions (push/pop) and the side menu effect. I am unsure about the shared element transitions between page-to-page, as that is a fairly complex implementation detail for a framework to create, that would work for all instances of how a developer's UI may be constructed. Likely in that scenario we would need to investigate if there is an API that could be exposed to enable developers an easier implementation of their own custom experience.

**Comment by jaytonic** [comment]
@sean-perkins Sure.
The most used components that I've seen differences:

- Global The colors and their variations are more subtle. All the "filled" options have using a variation of the main color
- Buttons: now should be more round, shadow is different
- Cards: are more rounded, shadow is different
- Toggle are now quite different(closer to what is done for iOS
- Floating action buttons are now squircle(not sure it's the exact term, but square with rounded borders), without shadows
- Input text have a background colors. They also have a slight border radius
- tabs: They now have an active indicator below
- chips: They are now more squared
- drawer: the drawer overlay has some border radius

No doubts there are other stuff, but those are the one we identified as blocking for our designs.

So basically, a lot of things around border-radius, shadows, shade of colors to be used

**Comment by laurentpayot** [comment]
Major changes are listed at https://material.io/blog/migrating-material-3

**Comment by martinzima** [comment]
@sean-perkins Besides the missing features that @jaytonic mentioned, I would also like to add [collapsible large titles](https://m3.material.io/components/top-app-bar/overview) as something that we miss greatly (also discussed in this [closed issue](https://github.com/ionic-team/ionic-framework/issues/25157)). I feel this is something that can affect the overall looks of an Android app a lot and the current design looks outdated.

**Comment by RudraSen2** [comment]
Material 3 Example - https://callstack.github.io/react-native-paper/

**Comment by RudraSen2** [comment]
Any ETA for this?

**Comment by johannesvollmer** [comment]
We're also depending on this. Especially the color design tokens, and generating the pallete (and modifying the pallette)

**Comment by danielkleebinder** [comment]
@johannesvollmer As for color generation, you can rely on the official material color utilities made by the material foundation:

- [GitHub material-foundation/material-color-utilities](https://github.com/material-foundation/material-color-utilities)
- [npm @material/material-color-utilities](https://www.npmjs.com/package/@material/material-color-utilities)

However, this package is incomplete with regards to some of the surface container colors, shadow tokens and additional custom colors. That's the reason, why I created a repository that adds all missing tokens:

 - https://github.com/danielkleebinder/md3-for-ionic

It also overrides some of the Ionic component styles, but this feature is still incomplete. Buttons and navigation bar, however, should work fine.

**Comment by johannesvollmer** [comment]
Thanks for the hint, we will check it out :)

**Comment by xnousnow** [comment]
It's been few years since this issue has been opened... Is there any updates here with ionic 7?

**Comment by danielkleebinder** [comment]
@sean-perkins Regarding shared element transitions, browsers now expose a new API (I suppose Google used its influence here) called the "View Transitions API". You can read on the details here:

- [Smooth and simple transitions with the View Transitions API (Chrome for Developers)](https://developer.chrome.com/docs/web-platform/view-transitions/)
- [View Transitions API (mdn web docs)](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API)
- [Animate between states with the View Transitions API (YouTube)](https://youtu.be/xZteiW-cun8?si=_M3ppcarBFqO4J8l)

It pretty much looks like it was made for Material Design 3!

**Comment by ozgurg** [comment]
Is there the source code of the Android screenshot shown in the "Infinitely Customizable" section on https://ionicframework.com/? It's very similar to Material 3.


![2024-01-02_02-55](https://github.com/ionic-team/ionic-framework/assets/6717356/aefadcf7-7d99-48c6-91b3-1b34155caf18)

**Comment by xnousnow** [comment]
I think they're finally working on it?

**Comment by omar2205** [comment]
> I think they're finally working on it? Also the team should check out https://m3.material.io/develop/web

The web version is deployed here https://material-web.dev/

**Comment by laurentpayot** [comment]
> Is there the source code of the Android screenshot shown in the "Infinitely Customizable" section on https://ionicframework.com/? It's very similar to Material 3.

@ozgurg [the same comment was made 10 months ago](https://github.com/ionic-team/ionic-framework/issues/24282#issuecomment-1546911285) and unfortunately nothing has moved since.

**Comment by ozgurg** [comment]
> > Is there the source code of the Android screenshot shown in the "Infinitely Customizable" section on https://ionicframework.com/? It's very similar to Material 3.
> 
> @ozgurg [the same comment was made 10 months ago](https://github.com/ionic-team/ionic-framework/issues/24282#issuecomment-1546911285) and unfortunately nothing has moved since.

Opps, my mistake, but it wouldn't hurt to ask again. :)

**Comment by danielkleebinder** [comment]
> Any ETA for this?

> It's been few years since this issue has been opened... Is there any updates here with ionic 7?

@sean-perkins Can you tell us a little bit about current state of Material Design 3 for Ionic? Are you coupling its implementation to the [official Material Design Web Components library](https://github.com/material-components/material-web)? Is MD3 planned for Ionic 8? Is there any way in which we can help you implementing these design guidelines?

**Comment by half-real-SCRACX** [comment]
Any updates? Official MD3 support would be essential for a lot of us. Or at least you could write some docs for ways to implement MD3 in Ionic through 3rd party frameworks.
Ionic is very useful, but it's a shame that by default it looks outdated on Android...

**Comment by xnousnow** [comment]
It's been a few years and it's still not here; is it even being worked on?

**Comment by half-real-SCRACX** [comment]
> Not a single component or progress update here; is it even being worked on?

Probably not... ☹️
But why? They said they're focussing on Ionic7, but 8 already came around and there still are no updates about Material3.

**Comment by italosll** [comment]
Any updates?

**Comment by xnousnow** [comment]
Ionic’s goal is to let you build web apps that look and feel native. But with almost three years since MD3’s release and no visible progress, it seems like it’s not being prioritized. That’s disappointing since staying current with native design is crucial and undermines Ionic’s purpose.

**Comment by tanishqmanuja** [comment]
This is my attempt to add MD3 to ionic using various other open source libraries. Sharing cause this might be helpful reference for others.

https://github.com/tanishqmanuja/demo.angular-ionic-material

**Comment by italosll** [comment]
@sean-perkins any updates on ionic supporting Material Design 3?

**Comment by severol** [comment]
@brandyscarney Sean seems to have moved to other interests, do you have any update or could you assign someone from @ionic-team to this?

**Comment by alexskrms** [comment]
Hi, 2025. Any news about this topic?

**Comment by NyaomiDEV** [comment]
if anything i'd say the collapsible headers are the hardest thing to achieve in ionic without a first party md3 implementation... everything else is "achievable".

i'm basically manually styling my app which uses ionic 8 if anyone is interested: https://github.com/NyaomiDEV/Ampersand/tree/main/src/lib/theme

truly homegrown and probably hard to extract to its own library but the css alone could be used by someone else i guess

**Comment by rdlabo** [comment]
The collapsible headers could not be implemented, but it was possible to make the headers disappear when scrolling.
Since the library is publicly available, Angular users can start using it immediately.

https://github.com/rdlabo-team/ionic-angular-library/tree/main/projects/scroll-header#readme

DEMO: https://rdlabo-ionic-angular-library.netlify.app/main/scroll-header

**Comment by NyaomiDEV** [comment]
@rdlabo the collapsible headers could be implemented, provided you almost copy the iOS implementation ionic already has

to be brief, you should place a component the full height of the expanded header you want in ion-content and setting said ion-content to fullscreen, then you could use intersection observer to get the overlap between the phantom header (scrollable) and the fixed header (non-scrollable but which we should animate somehow) and calculate the real header height and animation step from there

the problem I face is that properly animating this contraption is practically hell for my brain

**Comment by aeharding** [comment]
It hasn't been mentioned yet, but with md3 I think it would be really cool to be able to provide a custom user-specified colorway/tint:

https://m3.material.io/styles/color/dynamic/user-generated-source

Ionic could even take advantage of emerging CSS [color-mix()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) function so that the Ionic developer could specify a "tint" in hex, and it would just work™️ to dynamically theme the application - all from one hex code.

All these new CSS color functions could really simplify Ionic's existing theming, too: https://ionicframework.com/docs/theming/color-generator

**Comment by danielkleebinder** [comment]
Since no progress updates have been reported by the Ionic team, I have started implementing a small toolkit that adds [Material Design 3](http://m3.material.io/) to Ionic applications. It generates MD3 schemes for **light and dark** mode and overrides the default Ionic styles. The MD3 tokens are generated using the official [Material utilities library](https://github.com/material-foundation/material-color-utilities). You can find the toolkit here:

- [https://github.com/danielkleebinder/md3-for-ionic](https://github.com/danielkleebinder/md3-for-ionic)

Simply clone the repository, install the dependencies and run the program:

```bash
git clone https://github.com/danielkleebinder/md3-for-ionic.git
cd md3-for-ionic
npm install

node . --output="./out" --color=#ff0000 --with-ionic-overrides
```

Add the output files the the project you want MD3 to be applied to and add the following import to your `global.scss` file:

```css
@import "theme/md3/theme.css";
```

**You can find the Ionic kitchen sink demo with MD3 applied to it [here on Vercel](https://md3-for-ionic-demo.vercel.app/).**

_**Note:** I am going to maintain this project until official support is released, because I need MD3 for a couple of projects of mine. I cannot release MD2 apps to Android anymore. Please let me know if this toolkit is of any use to you._

**Comment by italosll** [comment]
> Since no progress updates have been reported by the Ionic team, I have started implementing a small toolkit that adds [Material Design 3](http://m3.material.io/) to Ionic applications. It generates MD3 schemes for **light and dark** mode and overrides the default Ionic styles. The MD3 tokens are generated using the official [Material utilities library](https://github.com/material-foundation/material-color-utilities). You can find the toolkit here:
> 
> * https://github.com/danielkleebinder/md3-for-ionic
> 
> Simply clone the repository, install the dependencies and run the program:
> 
> git clone https://github.com/danielkleebinder/md3-for-ionic.git
> cd md3-for-ionic
> npm install
> 
> node . --output="./out" --color=#ff0000 --with-ionic-overrides
> Add the output files the the project you want MD3 to be applied to and add the following import to your `global.scss` file:
> 
> @import "theme/md3/theme.css";
> **You can find the Ionic kitchen sink demo with MD3 applied to it [here on Vercel](https://md3-for-ionic-demo.vercel.app/).**
> 
> _**Note:** I am going to maintain this project until official support is released, because I need MD3 for a couple of projects of mine. I cannot release MD2 apps to Android anymore. Please let me know if this toolkit is of any use to you._

looks stunning, congratulations!

**Comment by laurentpayot** [comment]
By the way Material Design **3** is already so old that Google refreshed it with the ["Expressive" evolution](https://m3.material.io/blog/building-with-m3-expressive) 😅

**Comment by danielkleebinder** [comment]
@italosll Thank you very much! 😊

@laurentpayot It's truely remarkable how "fast" design is moving nowadays. I just recently noticed the design changes and adapted the library mentioned above to (at least partially) support the [spring motion physics system](https://m3.material.io/styles/motion/overview/how-it-works):

https://github.com/user-attachments/assets/c8e0aa81-779d-455b-bd17-6a7ad7f3b57e

**Comment by laurentpayot** [comment]
@danielkleebinder The View Transitions API is awesome, I use it with [Beercss](https://www.beercss.com/) to get a native-like experience with minimal bloat.

**Comment by Xefir** [comment]
@danielkleebinder I'm really impress by your work, good job !

I made a Capacitor plugin to integrate easily in three line of code your style library and sync it with the current color palette of the user on Android.

You can find my plugin in my Gitea : https://git.crystalyx.net/Xefir/ionic-md3

There are limitations due to how the CSS works, but in most case, it works great.

Have a nice day.

**Comment by 7freaks-otte** [comment]
While desperately waiting for MD3 and iOS26 design, I'll drop my approach on how I achieved collapsible headers `md` mode:

This depends on a boolean property. Up to you if you manually set that or replicate Ionic´s internal `getIonMode(ref?: HTMLElement) { ... }` function to detect it from the `ion-content` element:

```ts
ionMd3 = document.documentElement.getAttribute('mode') === 'md';
```

Then in your markup

```html
<ion-header>
    <ion-toolbar>
        <ion-title>Page title</ion-title>
    </ion-toolbar>
</ion-header>

<ion-content>
    <ion-header collapse="condense" mode="ios"><!-- must always be ios -->
        <ion-toolbar [attr.mode]="ionMd3 ? 'md' : null" [ngClass]="{ ios: !ionMd3, md: ionMd3 }">
            <ion-title>Page title</ion-title>
        </ion-toolbar>
    </ion-header>
...
</ion-content>
```

And the CSS:
```css
.header-md {
    box-shadow: none;
}
ion-content.md > .header-collapse-condense.ios {
    display: block;
    ion-title[size="large"] {
        position: absolute;
        bottom: 0;
        font-size: 34px;
    }
}
```

Don't ask my why exactly but I had to play around quite a bit to get that to work and I needed to use `attr.mode` instead of `[mode]` or `mode="{{}}"`

**Comment by gnbm** [comment]
Hey everyone👋🏻

We've seen the passion and long-standing requests for full Material Design 3 (MD3) support in the Ionic Framework! It's clear that the shift towards a more dynamic, personal, and modern design system is something you all want, and we're totally on board with that vision.
We appreciate you keeping this important conversation going.

However, due to our current strategic focus, a full native MD3 implementation is not on our immediate roadmap.
Here's why we're focusing our energy elsewhere:

**Ionic Framework v9's Modular Magic** ✨
Our top priority right now is Ionic Framework v9.
This release involves some massive architectural changes to make Ionic far more modular in terms of theming.
This modularity is the game-changer! It's designed specifically to make it incredibly easy for the community to create and share your own custom themes—including themes that perfectly capture the Material Design 3 look—without waiting on us!
We're building the foundation to empower you to design exactly what you want.

If you are curious about this, here’s the [branch](https://github.com/ionic-team/ionic-framework/tree/ionic-modular) where we’re currently working on it :nerd_face:

Thanks again for the input—keep the ideas coming!
